controllers/application_controller.go

```diff
@@ -0,0 +1,115 @@
+}
+
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=applications,verbs=get;list;watch;update
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=applications/finalizers,verbs=update
+
+func (r *ApplicationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.finalizers = finalizer.NewFinalizers()
+	if err := r.finalizers.Register(linkedRemoteSecretsTargetFinalizerName, &linkedRemoteSecretTargetsFinalizer{client: r.k8sClient}); err != nil {
+		return fmt.Errorf("failed to register the linked remote secret finalizer: %w", err)
+	}
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&appstudiov1alpha1.Application{}).
+		Complete(r)
+	if err != nil {
+		return fmt.Errorf("failed to configure the application reconciler: %w", err)
+	}
+	return nil
+}
+
+func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	lg := log.FromContext(ctx)
+	lg.V(logs.DebugLevel).Info("starting reconciliation")
+
+	var application appstudiov1alpha1.Application
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, &application); err != nil {
+		if errors.IsNotFound(err) {
+			lg.V(logs.DebugLevel).Info("Application already gone from the cluster. skipping reconciliation")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get the Application: %w", err)
+	}
+
+	finalizationResult, err := r.finalizers.Finalize(ctx, &application)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to finalize: %w", err)
+	}
+	if finalizationResult.Updated {
+		if err = r.k8sClient.Update(ctx, &application); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update application based on finalization result: %w", err)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+type linkedRemoteSecretTargetsFinalizer struct {
+	client client.Client
+}
+
+var _ finalizer.Finalizer = (*linkedRemoteSecretTargetsFinalizer)(nil)
+
+// Finalize removes the remote secret targets synced to the actual application which is being deleted
+func (f *linkedRemoteSecretTargetsFinalizer) Finalize(ctx context.Context, obj client.Object) (finalizer.Result, error) {
+	application, ok := obj.(*appstudiov1alpha1.Application)
+	if !ok {
+		return finalizer.Result{}, unexpectedObjectTypeError
+	}
+
+	remoteSecretsList := rapi.RemoteSecretList{}
+	if err := f.client.List(ctx, &remoteSecretsList, client.InNamespace(application.Namespace)); err != nil {
+		return finalizer.Result{}, unableToFetchRemoteSecretsError
+	}
+
+	if len(remoteSecretsList.Items) == 0 {
+		return finalizer.Result{}, nil
+	}
+
+	for rs := range remoteSecretsList.Items {
+		remoteSecret := remoteSecretsList.Items[rs]
+		applicationInSecret, set := remoteSecret.Labels[ApplicationLabelName]
+		if set && applicationInSecret != "" && applicationInSecret != application.Name {
+			// this secret is intended for another application, bypassing it
+			continue
+		}
+		// cleanup the targets
+		remoteSecret.Spec.Targets = []rapi.RemoteSecretTarget{}
```

Comment:
Wouldn't it be more logical to actually remote the remote secret? The application that it belongs to is being deleted, so what purpose would the RS serve after that? Do we want to keep it around just in case someone wanted to reuse it in another application? How would that work?

Reply:
I was thinking of this, the only concern was the accidental loss of secret data, but i thing we can go with it unless anyone complain :)

Reply:
Well, I think we should be cautious here and actually care more about accidental secret data retention, which is what happens here - these remote secret just hold on to data and are no longer reachable from anywhere.